### PR TITLE
CLN: Preliminary formatter refactor

### PR DESCRIPTION
**Pull request by shangyian**
- [x] based on discussion here: https://github.com/pandas-dev/pandas/pull/20032#pullrequestreview-101917292
- [x] tests added / passed
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [ ] whatsnew entry

This PR splits the various formatters into different files so that it's easier to refactor a specific formatter.

### CODE REVIEW

File: pandas/io/formats/common.py [hunk 1]
```diff
@@ -42,3 +45,36 @@ def get_level_lengths(levels, sentinel=''):
         result.append(lengths)
 
     return result
+
+
+def buffer_put_lines(buf, lines):
+    """
+    Appends lines to a buffer
```
Comment (jreback) [inline pandas/io/formats/common.py]: can you use numpy doc string format here

File: pandas/io/formats/latex.py [hunk 2]
```diff
@@ -0,0 +1,244 @@
...
+    Parameters
+    ----------
+    formatter : `DataFrameFormatter`
+    column_format : str, default None
+        The columns format as specified in `LaTeX table format
+        <https://en.wikibooks.org/wiki/LaTeX/Tables>`__ e.g 'rcl' for 3 columns
+    longtable : boolean, default False
+        Use a longtable environment instead of tabular.
+
+    See also
```
Comment (jreback) [inline pandas/io/formats/latex.py]: See Also

File: pandas/io/formats/html.py [hunk 3]
```diff
@@ -0,0 +1,505 @@
...
+from textwrap import dedent
+
+import pandas.core.common as com
+from pandas.core.index import MultiIndex
+from pandas import compat
+from pandas.compat import (lzip, range, map, zip, u,
+                           OrderedDict, unichr)
+from pandas.core.config import get_option
+from pandas.io.formats.printing import pprint_thing
+from pandas.io.formats.common import get_level_lengths, \
```
Comment (jreback) [inline pandas/io/formats/html.py]: use parens rather than line continuation ``\``

File: pandas/io/formats/csv.py [hunk 4]
```diff
@@ -0,0 +1,280 @@
...
+from pandas import compat
+from pandas.compat import (StringIO, range, zip)
+
+from pandas.io.common import (_get_handle, UnicodeWriter, _expand_user,
+                              _stringify_path)
+from pandas._libs import writers as libwriters
+from pandas.core.indexes.datetimes import DatetimeIndex
+from pandas.core.indexes.period import PeriodIndex
+import numpy as np
+
```
Comment (jreback) [inline pandas/io/formats/csv.py]: put these import first (csv then numpy)

**Review by jreback — CHANGES_REQUESTED** [review]
looks good. just some very minor comments. want to do this refactor w/o really changing much code (as its more transparent that way). next pass will clean / fix more things.

File: pandas/io/formats/csv.py (hunk 4 above)
Comment (shangyian) [reply pandas/io/formats/csv.py]: Cool, just fixed these

**Comment by shangyian** [comment]
Hmm. Not sure why Travis failed and AppVeyor succeeded.

**Comment by jreback** [comment]
you have a lint error
```
inside ci/lint.sh
Linting *.py
pandas/io/formats/format.py:33:1: F401 'pandas.io.formats.csv_format.CSVFormatter' imported but unused
Linting *.py DONE
```

**Comment by jreback** [comment]
ok I fixed up the imports a bit to make them more idiomatic. I think its possible to completely remove pandas.io.formats.common and just move these 2 functions back to .formats. No real reason to have an extra file here. see if you can do that.

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
thanks @shangyian 

will merge a bit later (after doc sprint)

**Comment by shangyian** [comment]
@jreback - cool. I also moved the two functions in io.formats.common to formats.format

File: pandas/core/frame.py [hunk 5]
```diff
@@ -1841,7 +1842,7 @@ def info(self, verbose=None, buf=None, max_cols=None, memory_usage=None,
             - If False, never show counts.
 
         """
-        from pandas.io.formats.format import _put_lines
+        from pandas.io.formats.format import buffer_put_lines
```
Comment (jreback) [inline pandas/core/frame.py]: you can reference this directly as ``fmt.buffer_put_lines`` as fmt is already imported

File: pandas/core/frame.py [hunk 6]
```diff
@@ -1940,7 +1941,7 @@ def _sizeof_fmt(num, size_qualifier):
             mem_usage = self.memory_usage(index=True, deep=deep).sum()
             lines.append("memory usage: %s\n" %
                          _sizeof_fmt(mem_usage, size_qualifier))
-        _put_lines(buf, lines)
+        buffer_put_lines(buf, lines)
```
Comment (jreback) [inline pandas/core/frame.py]: same

**Review by jreback — CHANGES_REQUESTED** [review]
great. there is also a ``_get_level_lengths`` in ``style.py``, you *might* be able to combine these (can do here or next PR)

File: pandas/core/frame.py (hunk 6 above)
Comment (shangyian) [reply pandas/core/frame.py]: Good point, changed those. I'll take a look at the two `get_level_lengths`, but may not get around to fixing that today, so it's probably better as a new PR

**Comment by jreback** [comment]
can you rebase on master. ping on green.

**Pull request merged**